src/shogun/clustering/KMeansBase.cpp

```diff
@@ -130,6 +131,115 @@ void KMeansBase::compute_cluster_variances()
 	}
 }
 
+SGMatrix<float64_t> KMeansBase::compute_std_dev() const
+{
+    require(cluster_centers.size() > 0, "KMeans is not trained!");
+
+    auto lhs_cache = distance->get_lhs();
+    distance->replace_lhs(distance->get_rhs());
+
+	SGMatrix<float64_t> stds(dimensions, k);
+
+	auto rhs = distance->get_rhs()->as<DenseFeatures<float64_t>>();
+	SGVector<int32_t> cluster_assignments;
+	SGVector<int64_t> weights_set;
+	std::tie(cluster_assignments, weights_set, std::ignore) =
+	    compute_cluster_assignments(k);
+
+	auto cluster_indexes = std::vector<SGVector<index_t>>(k);
+	auto cluster_counters = SGVector<index_t>(k);
+	for (int32_t current_cluster = 0; current_cluster < k; ++current_cluster)
+	{
+		cluster_indexes[current_cluster] =
+		    SGVector<index_t>(weights_set[current_cluster]);
+		cluster_counters[current_cluster] = 0;
+	}
+
+	for (int32_t i = 0; i < rhs->get_num_vectors(); ++i)
+	{
+		int point_cluster = cluster_assignments[i];
+		int& cluster_counter = cluster_counters[point_cluster];
+		cluster_indexes[point_cluster][cluster_counter] = i;
+		++cluster_counter;
+	}
+
+	for (int32_t i = 0; i < k; ++i)
+	{
+		stds.set_column(i, view(rhs, cluster_indexes[i])->std());
```

Comment:
yep this is what I had in mind, nice.
The only thing is, I just checked the `std` implementation, and it only produces valid results if no subset is present (it is missing an assert for that atm).
Could you add an if-else branch for the case where subsets are present, that iterates through the feature vectors one by one to compute the std? If possible, please iterate in a memory friendly order (sorting the subset indices first)

@gf712 didnt we have something like this automagically somewhere?

Reply:
as this implementation here is wrong, this shows that we need some unit tests for the stds :)

Reply:
I'd also prefer if the std data wouldnt be copied here, but the std() method would write it straight to the matrix. E.g. via first getting the colunm vector reference, and then passing that to std or something

Reply:
We have an iterator for features, but the subset is not sorted as it took longer to qsort than the cheapest computations. I would say test with manual sorting and with iterator.

Reply:
I thought about this a bit more. The solution here is actually not ideal. You go through all the data and identify the cluster centers corresponding to each point. Then you copy the data for each cluster and compute the std. While the copying can be avoided, we can turn the whole approach upside down to make it simpler (no need to compute this data structure you added) and more efficient: just compute the cluster for each training point (i.e. apply method or similar), then just iterate through the data *once*, and add the contribution for the std deviation of each point to the corresponding cluster std

```
clusters = kmeans.apply(features) # vector with cluster inds
std_devs = SGMatrix(num_dims, num_data) # init with zero
for (vec, c_idx) in zip(feats, clusters):
  mean = m_cluster_centres[c_idx] # assumes this was computed and stored by "train"
  std_devs[:,c] += (vec-mean)**2 # this is naive (unstable), you want to use Welford's alg or similar, see [here](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance)
std_devs /= (num_vec-1)
std_devs = linalg::sqrt(std_devs)
```

Reply:
We are calculating std for each cluster thus we need to know which points belong to this cluster. We can store cluster assignments during Lloyd K-Means, but it is not so easy in the MiniBatch case.

Reply:
https://github.com/shogun-toolbox/shogun/blob/63147ff466b5ce5e51966d2879318e57e7e78d46/src/shogun/clustering/KMeansMiniBatch.cpp#L79
The only way is to add here something like:
```
cluster_assginments[M[j]] = near;
```

Reply:
couldnt you run the `apply` function to get the cluster assignments of the training data? Or some helper function defined within?

Reply:
for mini batch we would need to compute it on the fly I guess?

Reply:
Thank you! I did not know about it. Now I can use _apply_ in Lloyd and in Minibatch to get assignments simply adding three lines of code. So extracting _compute_cluster_assignments_ was a bad idea from the outset.